src/lib.rs

```diff
@@ -134,10 +197,11 @@ impl<T> Storage<T> {
         for index in pending.add_ref.drain(..) {
             s.meta[index] += 1;
         }
-        for index in pending.sub_ref.drain(..) {
+        let (mut refs, epoch) = pending.prepare_sub_drain();
+        for index in refs.drain(..) {
             s.meta[index] -= 1;
             if s.meta[index] == 0 {
-                s.free_list.push(index);
+                s.free_list.push((index, epoch[index] + 1));
```

Comment:
I think we actually need to bump the `epoch` here, as in `epoch[index] += 1`, unless I'm missing the place where you are already doing that.